File: java/src/main/java/codingdojo/WordWrapBlausten2.java
package codingdojo;

/**
 * This code is adapted from https://github.com/sblausten/word-wrap-kata
 */
public class WordWrapBlausten2 implements WordWrap {
    @Override
    public String wrap(String line, int maxLineLen) {
        return WordWrapBlausten2.wrap_static(line, maxLineLen);
    }

    private static String wrap_static(String text, int cols) {
        if (cols >= text.length()) return text;
        int colsToSpace = text.substring(0, cols).lastIndexOf(' ');
        int actualCols = colsToSpace == -1 ? cols : colsToSpace + 1;
        if (text.charAt(cols) == ' ') actualCols = cols + 1;
        return (text.substring(0, actualCols) + "\n" + WordWrapBlausten2.wrap_static(text.substring(actualCols), cols));
    }
}

File: typescript/src/WordWrapBlausten2.ts
import WordWrap from "./WordWrap";

export class WordWrapBlausten2 implements WordWrap {
    public wrap(line: string, maxLineLen: number): string {
        return wrap_static(line, maxLineLen);
    }
}

function wrap_static(text: string, cols: number): string {
    if (cols >= text.length) return text;
    const colsToSpace = text.substring(0, cols).lastIndexOf(' ');
    let actualCols = colsToSpace == -1 ? cols : colsToSpace + 1;
    if (text.charAt(cols) == ' ') actualCols = cols + 1;
    return (text.substring(0, actualCols) + "\n" + wrap_static(text.substring(actualCols), cols));
}